flink-runtime/src/test/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContextTest.java

```diff
@@ -170,6 +178,70 @@ public void testSnapshotAndRestore() throws Exception {
 
 	}
 
+	@Test
+	public void testCallableInterruptedDuringShutdownDoNotFailJob() throws InterruptedException {
```

Comment:
This test looks to be testing against a specific implementation, rather against a contract.
You need to assume the exact specific way of how the `close()` method is implemented for this test to be meaningful. The test re-engineers how `close()` works to produce a special race.

That usually causes issues: as soon as the `close()` implementation changes
  - either the test breaks (despite everything being okay)
  - or the test becomes meaningless (doesn't reproduce the critical situation any more).

Testing such concurrent race situations is always a bit tricky. I think the "manual executors" have a decent trade-off between reliably producing situations and not assuming too much implementation.
You could take a look at the `ContinuousFileSplitEnumeratorTest` and the `ManuallyTriggeredScheduledExecutorService` in the `TestingSplitEnumeratorContext`.

Using that, you can probably replace all the complex code in the test by:
```java
// async action before the enumerator closes
testingContext.callAsync(actionThatThrowsException);
asyncExecutor.triggerAll();

// callback after the enumerator closes
testingContext.close();
enumeratorExecutor.triggerAll();

assertFalse(operatorCoordinatorContext.isJobFailed());
```

This only ties the test to the implementation aspect that there are two executors, not any specific structure of the close method.